Async Refactor WIP

fel115 opened the pull request.

horenso commented:
Seems like it only needs formatting.

horenso commented:
We can merge this to have one consistent "old" version before the complete refactor.

markus2330 commented:
> We can merge this to have one consistent "old" version before the complete refactor.

If it is tested on hardware that it still works, we can do so, otherwise I don't see why we should do it. Master should be known to work. You can simply give `0.8.*` tags on your branch when something is consistent. When we finish this refactoring, we can open 0.9.* for testing/fixing (then again on master) and finally we can release 1.0.0.

markus2330 commented on files/opensesame.spec:
```diff
@@ -126,6 +126,11 @@ description=How often to get new data (default: every 60 seconds, which is the h
 default = 6000
 type = unsigned_short
 
+[sensors/enable]
```
What happens if the sensors get started with no sensors available? To avoid it, I used the check if there is at least one element in the array. But I am also okay with `sensors/enable` for consistency reasons.

fel115 replied:
https://github.com/ElektraInitiative/opensesame/blob/5ede5a8dc4a27e5cd733353df23f727fd7485a8e/src/sensors.rs#L147-L150
Do you mean this lines ?

markus2330 commented on src/main.rs:
```diff
@@ -635,7 +866,88 @@ fn normal_mode(
+		nextcloud_receiver,
+	));
+
+	if garage_enabled {
+		if !buttons_enabled {
+			panic!("Garage depends on buttons!");
+		}
+		tokio::spawn(garage_loop(
+			config,
+			command_sender.clone(),
+			nextcloud_sender.clone(),
+		));
+	}
```
As said, I would prefer if modules are self-contained, all logic/code related to a module should be within a module.

markus2330 replied:
Yeah, actually we could simply have one function as interface to the module: the loop. I was thinking more of having `handle` as entry point, but why not, the whole loop is even more flexible.

> Or we spawn a task for every module and if an error occurs it leaves the task. And the error code could be something like NotEnabled. 

Somehow the main framework needs to be noticed about modules leaving their main loop in error states (different than to leaving it in disabled state).

markus2330 commented on src/mod_ir_temp.rs:
```diff
@@ -45,7 +45,7 @@ impl ModIR {
 		Self {
 			mlx: None,
-			device: "/dev/null".to_string(),
+			device: String::from("/dev/null"),
```
As Elektra already provides defaults, it is more elegant if there are not other defaults in the source code.

fel115 replied:
Yes. I think we are going to remove this function and handle it in the `new()` function.

markus2330 commented on src/todo:
```diff
@@ -0,0 +1,31 @@
+button module
+
+alle 10ms: schauen ob ein button gedrückt wurde
```
You could also try with a bit more than 10ms, as currently it is also 10ms delay+time to execute the logic.

markus2330 commented on src/todo:
```diff
@@ -0,0 +1,31 @@
+
+alle 10ms: schauen ob ein button gedrückt wurde
+
+BUTTON                       VALIDATER
```
Note that validator is not a module on its own but code used within the module buttons

```suggestion
BUTTON                       VALIDATOR
```

markus2330 reviewed (commented):
Good to see progress, cannot wait to see more :rocket:

horenso commented:
@markus2330 I have a few questions regarding config. Can we just have multiple copies of config cloating around, or should there only be one config object that is atomic? what about sync, is that done by libelectra?

horenso commented:
@markus2330 I have a few questions regarding config. Can we just have multiple copies of config cloating around, or should there only be one config object that is atomic? what about sync, is that done by libelectra?

markus2330 commented:
> I have a few questions regarding config. Can we just have multiple copies of config cloating around,

Elektra's core definitely allows it but there are some plugins in Elektra that don't like concurrency. E.g. https://github.com/ElektraInitiative/libelektra/issues/4466 was a problem for Opensesame but it should be fixed now. Furthermore, there might be issues if there are way too many open handles: https://github.com/ElektraInitiative/libelektra/issues/3140

> or should there only be one config object that is atomic?

Yeah, some mutex protection or similar is probably the simpler and safer way to use Elektra.

> what about sync, is that done by libelectra?

Which sync do you mean? Sync on file system level is done by libelektra.

markus2330 commented:
> Elektra's core definitely allows it

To clarify: it is allowed (with potential limitations described above) to have separated instances of KDB, Key and KeySet per thread and use them concurrently.

fel115 commented:
@horenso I have implemented the signal module, by using `config` and `state` as `Arc<Mutex<Config<'a>>>`. But the problem with Config is, that it doesn't implement `Send`, so the problem is that we have 3 modules that need to run with `spawn_local` (weatherstation, signals, environment). Maybe it is okay to run these three modules in one Thread, but we also need to figure out how the `spawn_local` function works. 

I tested it on the weather station and the `async fn get_background_task()` didn't do anything. I added a `println!` to the beginning of this function and I didn't get a response. And `weatherstation` was the only module running!

markus2330 commented on src/audio.rs:
```diff
@@ -38,24 +49,40 @@ impl Audio {
+						let ssh_result =
+							exec_ssh_command(String::from("killall -SIGUSR2 opensesame")).await;
+						if let Err(err) = ssh_result {
+							eprintln!(
```
@horenso all logging should be done to the ping channel of Nextcloud.

fel115 replied:
Is fixed with commit [4a8d992](https://github.com/ElektraInitiative/opensesame/pull/75/commits/4a8d9923af7c4b6be8729fda63bae5d4638412c5).

fel115 commented:
I think the refactoring to async is done. Other changes can be done in a separate PR. I already tested Buttons, Bell, PWR, Environment, Battery, Watchdog, Audio Bell, Garage, Signal, and Sensors. Today I am going to test Weatherstation in combination with other modules. 

@horenso @markus2330

markus2330 commented:
> I think the refactoring to async is done. Other changes can be done in a separate PR. I already tested Buttons, Bell, PWR, Environment, Battery, Watchdog, Audio Bell, Garage, Signal, and Sensors.

Great! :rocket: 

Is it already ready to review? How to start reviewing this big PR, can you write some intro to the module system and how everything works together?

> Today I am going to test Weatherstation in combination with other modules.

Please watch out that the weatherstation keeps on publishing most of the time. (If you install a broken version, please replace it with a working version again afterwards.)

> Which version of opensesame should this PR be? v0.8.0?

Yes, at least if already everything works. Maybe even 0.9.0 to indicate we will be soon at 1.0.0?

fel115 commented:
> Is it already ready to review? How to start reviewing this big PR, can you write some intro to the module system and how everything works together?

Yes, now it's ready for review. I hope my explanation below is helping.

# OpenSesame

## audio.rs
This module handles audio output for playing fire alarms and bell sounds. This module can receive commands from the Module Buttons (Bell), Environment (FireAlarm), Nextcloud (FireAlarm, Bell), and Signals (FireAlarm, Bell).

## bat.rs
Checks the battery capacity every ten minutes and outputs to Nextcloud if it falls below 50%. If it goes below 50%, the threshold for the next Nextcloud message is set to 40%.

## buttons.rs
This module implements the `do_reset` if an error occurs on the MOD-IO2 modules. In the `async get_background_task`, we implemented how the buttons were handled in the old main, including the validator and command receiver. The command receiver receives commands like `opendoor` (used by Garage, Nextcloud), `ring_bell` (used by Environment, Signals), `switchlights` (used by Garage, Nextcloud), and `RingBellAlarm` (used by Signals).

## clima_sensor_us.rs 
This module works independently, sending warnings to Nextcloud and publishing to opensensemap. We needed to implement `Send` to use libmodbus with async functions.

## config.rs
No changes were made to this module.

## environment.rs
Changed the functions `rememberBaseline`, `restoreBaseline`, and added a `Muext` of state. We moved `handle_environment` from the main into the `get_background_task`. This module receives commands (RestoreBaseline, RememberBaseline) from Signals.

## garage.rs
This module is checked at intervals of 10 milliseconds, triggering Nextcloud Chat or Buttons commands if the button is pressed. Future changes will involve removing the interval and implementing trigger-oriented events because GPIO pins from the Olimex board are used, along with interrupts.

## mod_ir_temp.rs
This module is triggered at given intervals and warns by sending Nextcloud messages.

## nextcloud.rs 
Implements two loops: one for sending (`message_sender_loop`) messages and status to Nextcloud, and the other for receiving (`command_loop`) messages/commands from Nextcloud. Commands can be sent via Nextcloud chat by typing "\opensesame" to open the door, or other commands like "\ring_bell", "\fire_alarm", "\status", and "\switchlights true true".

## ping.rs
This module sends a ping message to Nextcloud if it receives the `SendPing` event. Other functions update `Env`, `EnvStatus`, `EnvError`, and `BatCapacity`, but these commands aren't used yet. They need to be triggered at the right spot in Environment and Battery to keep the ping up to date.

## pwr.rs 
No significant changes were made to this module.

## sensors.rs 
No major changes, only the `get_background_task` was added with a similar implementation as in the old version.

## signals.rs 
This module listens to system signals and executes the same events as in the old version.

## ssh.rs 
Changed to an async function.

## types.rs
This module contains the error types defined so far.

## validator.rs
No changes were made to this module.

## watchdog.rs
This background task is triggered every few seconds and writes to the specified file. The `test_get_background_task` function is used to simulate the watchdog stopping writing to the watchdog file.

## main.rs
In this loop, we initialize every module if it is enabled. Additionally, the MPSC (Multiple Producer, Single Consumer) channels are initialized here. To configure each module without encountering issues related to multiple access to one config object, we read the configuration in the main.

markus2330 commented on emu_dev/emulate.sh:
please move these two to sub-folder in files

markus2330 commented on files/opensesame.spec:
```diff
@@ -54,6 +54,11 @@ description = which chat to use for sending ping messages. Note: this chat will
 check/length/max = 8
 
+[nextcloud/chat/commands]
+description = which chat is used to send commands to Opensesame.
```
Can it be the same chat as for sending? Can this be the default? (Then required wouldn't be needed)

markus2330 commented on files/opensesame.spec:
```diff
@@ -161,6 +179,13 @@ type = unsigned_short
+
+[sensors/log]
+description = logging sensor data to file
+default = "/home/olimex/data.log"
```
Does /dev/null work?

```suggestion
default = "/home/olimex/data.log"
example = "/dev/null"
```

fel115 replied:
We removed `sensors/log` because you said we shouldn't write to the SD card. Do you think we should implement the logging?

markus2330 replied:
Then please also remove from spec.

> Do you think we should implement the logging?

Can also be implemented later, not relevant for #75.

markus2330 commented on src/audio.rs:
```diff
@@ -0,0 +1,120 @@
+use futures::never::Never;
```
Please add a very short description in the top of the file and a longer description in the doc folder.

fel115 replied:
Issue #99

markus2330 commented on src/audio.rs:
```diff
@@ -0,0 +1,120 @@
+					nextcloud_sender
+						.send(NextcloudEvent::Chat(
+							NextcloudChat::Default,
+							gettext("🔔 Rining the Audio Bell "),
```
```suggestion
							gettext("🔔 Ringing the Audio Bell "),
```

markus2330 commented on src/audio.rs:
```diff
@@ -0,0 +1,120 @@
+				maybe_cancellation_token.unwrap().cancel();
+			};
+			maybe_cancellation_token = Option::Some(CancellationToken::new());
+			eprintln!("Cancelling previous audio...");
```
Logs only to Nextcloud. Probably remove this one.

markus2330 commented on src/bat.rs:
```diff
@@ -1,21 +1,78 @@
-			Ok(str) => return str.trim_end().parse::<u8>().unwrap(),
-			Err(_err) => return 100,
+			Ok(str) => str.trim_end().parse::<u8>().unwrap(),
+			Err(_err) => 0, // TODO: Do not return full bat in error case!
```
Please create issue or fix.

markus2330 commented on src/buttons.rs:
```diff
@@ -447,7 +478,233 @@ impl Buttons {
+					Buttons::do_reset(nextcloud_sender.clone(), &mut pwr).await?;
+				}
+			}
+			// Validation benötigt button, somit threads abhängig!!!; channel zwischen buttons und validator? damit validator nur getriggert ist wenn buttons sich ändert?
```
Please translate to English. Is this a TODO?

fel115 replied:
created an issue for that #92

markus2330 commented on src/main.rs:
```diff
@@ -1,670 +1,246 @@
-		nc.ping(gettext("👋 Turned PWR_SWITCH off"));
-		watchdog.trigger();
-		thread::sleep(time::Duration::from_millis(watchdog::SAFE_TIMEOUT));
+	let mut config = Config::new(CONFIG_PARENT);
+	let config_mutex = Arc::new(Mutex::new(Config::new(CONFIG_PARENT)));
+	let state_mutex = Arc::new(Mutex::new(Config::new(STATE_PARENT)));
```
Wouldn't it be more pretty if the mutex are within config?

fel115 replied:
Yes, I think that would be much more pretty. I try to solve that.

fel115 replied:
I created issue #94, but I think it would be better to implement it after merging this branch.

markus2330 replied:
Yes, it is also not super-important but rather cosmetic.

markus2330 commented on src/main.rs:
```diff
@@ -1,670 +1,246 @@
-			environment.to_string()
-		)),
+	if buttons_enabled {
+		let time_format = config.get::<String>("nextcloud/format/time");
+		let location_latitude = config.get::<f64>("location/latitude");
+		let location_longitude = config.get::<f64>("location/longitude");
+		tasks.push(spawn(Buttons::get_background_task(
+			Buttons::new(&mut config),
+			Validator::new(&mut config),
```
Why partially get a few configs but still pass config along? Why not pass along completely?

fel115 replied:
Yes, I think that would be also prettier. We did it that way because we had problems with creating a reference of config for every `get_background_task` function. But we could extend the struct of the Modules and save the data from the config in the structs of the modules.

fel115 replied:
Issue #96

markus2330 commented on src/main.rs:
```diff
@@ -1,670 +1,246 @@
-						format!("{:?}", sequence)
-					));
-				}
+	// if env_enabled || buttons_enabled {
```
Why is this commented out? Shouldn't there be a check if audio is enabled?

fel115 replied:
The audio module is enabled implicitly, but I think we should change it to an explicit one like in the other modules.

fel115 replied:
Issue #97

markus2330 commented on src/main.rs:
```diff
@@ -1,670 +1,246 @@
+))).await?;
+
+	join_all(tasks).await;
+	println!("hallo");
```
Write via Nextcloud.

fel115 replied:
We have already a Hello for the Nextcloud-Chat, this was only for testing. Will be removed.

markus2330 commented on src/mod_ir_temp.rs:
```diff
@@ -85,7 +72,16 @@ impl ModIR {
 				}
 			}
 		} else {
-			s = ModIR::new_default();
+			s = Self {
+				mlx: None,
+				device: String::from("/dev/null"),
+				addr: SlaveAddr::Default,
+				ambient_temp: 0.0,
+				object_temp: 0.0,
+				_emissivity: 1.0,
+				active_ambient_state: IrTempState::Normal,
+				active_object_state: IrTempState::Normal,
```
Is this else branch needed? (Can new be executed if there isn't a config? Why allow it?)

markus2330 commented on src/nextcloud.rs:
```diff
@@ -107,48 +135,227 @@ impl Nextcloud {
+							.iter()
+							.map(|m| m["message"].as_str().unwrap())
+						{
+							if message.starts_with('\\') {
```
This code needs some refactoring. Maybe together with leftovers (setpin etc.). Please create an issue. Unclear what already should work, please write this in documentation.

fel115 replied:
@horenso  I think you know it better than me, can you do that?

markus2330 replied:
Probably best via follow-up.

fel115 replied:
Issue #98

markus2330 commented on src/signals.rs:
```diff
@@ -0,0 +1,220 @@
+	}
+
+	async fn sigterm(&mut self) -> Result<(), ModuleError> {
+		//send message to environment to remember_baseline ans ENV need to use a Mutex of State
```
Seems like a TODO not marked as TODO.

fel115 replied:
created #93

markus2330 replied:
still not marked as TODO, and please also refer to #93 from the source.

markus2330 commented on src/signals.rs:
```diff
@@ -0,0 +1,220 @@
+			.send(NextcloudEvent::Chat(
+				NextcloudChat::Ping,
+				gettext!(
+					"👋reloading config&state for opensesame {} {}",
```
```suggestion
					"👋 reloading config&state for opensesame {} {}",
```

markus2330 commented on src/signals.rs:
```diff
@@ -0,0 +1,220 @@
+				.send(CommandToButtons::RingBellAlarm(20))
+				.await?;
+		}
+		//Wird mit play audio modul implementiert
```
Please consistently English.

markus2330 reviewed (changes requested):
Great progress, I like many of these changes. The documentation is a bit poor, please describe what works now for all the modules. There are a few German comments. Some TODOs are not clear. TODOs shouldn't be bugs in code (for that please create issues) but only things that could be improved.

markus2330 commented on src/sensors.rs:
```diff
@@ -244,7 +249,50 @@ impl Sensors {
+							gettext!("Fire Chat {}", w),
+						))
+						.await?;
+					println!("Chat - Sensors - {}", line);
```
There are still a few println not removed.

markus2330 reviewed (commented):
(only a few answers, did not review again)

markus2330 commented on src/sensors.rs:
```diff
@@ -244,7 +249,49 @@ impl Sensors {
+							gettext!("Fire Alarm {}", w),
+						))
+						.await?;
+					/*let mut state = state_mutex.lock().await;
```
Please mark with TODO+reference to issue what is to be done.

fel115 replied:
Issue #103

markus2330 reviewed (approved):
A few last issues are to be created and a few TODOs should refer to these issues, then we can merge this. I could test this tomorrow if we manage to merge it till then.

fel115 commented on doc/Testreport_Async_Refactor.md:
````diff
@@ -0,0 +1,122 @@
+watchdog.enable = 0 
+ping.enable = 0
+```
+### (`-`): This module doesn't enter the async-loop, because of problems with `spawn_local`
````
this is an old problem, know it is working. Will be changed in the next commit.

fel115 commented on doc/Testreport_Async_Refactor.md:
````diff
@@ -0,0 +1,122 @@
+done
+```
+### (`+`): Chat and Alarm trigger worked as expected
+### (`~`): **!!! implementation of state mutex is missing in async function of sensors !!!**
````
It is more a notification for the developers. I am going to create an issue #94.

fel115 commented on doc/Testreport_Async_Refactor.md:
````diff
@@ -0,0 +1,122 @@
+...
+```
+### (`+`): light button; light switch; light permanent; light time extended; pin validation
+### (`~`): First entering of pin caused time `sequence timeout`, the timeout was too fast. Besides this one failure it worked fine. Cloud be a startup issue.
````
Issue #102

fel115 commented:
I merged this branch with the main, and it is now ready.

markus2330 commented:
Finally done :rocket: :heart:
